sdk/core/src/main/java/org/apache/mesos/offer/constrain/AgentRule.java

```diff
@@ -34,52 +40,162 @@ public String toString() {
         return String.format("AgentRule{agentId=%s}", agentId);
     }
 
+    @Override
+    public boolean equals(Object o) {
+        return EqualsBuilder.reflectionEquals(this, o);
+    }
+
+    @Override
+    public int hashCode() {
+        return HashCodeBuilder.reflectionHashCode(this);
+    }
+
     /**
      * Ensures that the given Offer is located on the specified agent ID.
      */
+    @JsonIgnoreProperties(PassthroughGenerator.RULE_NAME) // don't include in serialization
     public static class RequireAgentGenerator extends PassthroughGenerator {
 
-        public RequireAgentGenerator(String agentId) {
+        private final String agentId;
+
+        @JsonCreator
+        public RequireAgentGenerator(@JsonProperty("agent_id") String agentId) {
             super(new AgentRule(agentId));
+            this.agentId = agentId;
+        }
+
+        @JsonProperty("agent_id")
+        private String getAgentId() {
+            return agentId;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("RequireAgentGenerator{agentId=%s}", agentId);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return EqualsBuilder.reflectionEquals(this, o);
+        }
+
+        @Override
+        public int hashCode() {
+            return HashCodeBuilder.reflectionHashCode(this);
         }
     }
 
     /**
      * Ensures that the given Offer is NOT located on the specified agent ID.
      */
+    @JsonIgnoreProperties(PassthroughGenerator.RULE_NAME) // don't include in serialization
     public static class AvoidAgentGenerator extends PassthroughGenerator {
 
-        public AvoidAgentGenerator(String agentId) {
+        private final String agentId;
+
+        @JsonCreator
+        public AvoidAgentGenerator(@JsonProperty("agent_id") String agentId) {
```

Comment:
Now might be a good opportunity to address this issue: https://mesosphere.atlassian.net/browse/INFINITY-637